### PR TITLE
Add count for current users online

### PR DESCRIPTION
**Pull request by lcezermf**
## Why is this change being made?

Adds #79 

## How is this being accomplished?

It adds a counter based on the amount of users accessing the website. This was done using Presence from Phoenix to handle the whole behavior.

Every LV module on the mount will trigger the Presence and a component on the home page will render the proper value.

Screenshot from 2024-07-22 18-21-16

## Testing

- Run the project
- Access the home page
- [ ] Must show the value of total online users
- Open a anon browser and access a new page
- [ ] It must update the value for the count

## Take aways

- UI is not my strongest skill so any suggestion or commit would be helpful
- There is a couple of copy and paste code, I do not see as a problem now, but in future that can be improved

### CODE REVIEW

File: lib/techschool_web/components/online_users.ex [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        <%= @count %> online users
+      </button>
+      <span class="flex absolute h-3 w-3 top-0 right-0 -mt-1 -mr-1">
+        <span class="animate-ping absolute inline-flex h-full w-full rounded-full bg-green opacity-75">
+        </span>
+        <span class="relative inline-flex rounded-full h-3 w-3 bg-green"></span>
+      </span>
+    </span>
+    """
+  end
```
Comment (danielbergholz) [inline lib/techschool_web/components/online_users.ex]: Some suggestions to simplify the HTML:

- Remove the edge case of checking for 1 user. We can add 1 if else inside the general case
- Convert the edge case of 0 users to 1 line
- I also refactored the UI a little bit and made the HTML more accessible

```suggestion
  def show_online_users(%{count: 0} = assigns), do: ~H""

  def show_online_users(assigns) do
    ~H"""
    <div
      id="online_users_count"
      class="mt-2 relative inline-flex"
      aria-label="Number of online users using TechSchool"
    >
      <p class="px-4 py-2 font-semibold text-sm rounded-md text-green bg-slate-800">
        <%= @count %> user<%= if @count == 1, do: "", else: "s" %> online
      </p>
      <span class="flex absolute h-3 w-3 top-0 right-0 -mt-1 -mr-1">
        <span class="animate-ping absolute inline-flex h-full w-full rounded-full bg-green opacity-75">
        </span>
        <span class="relative inline-flex rounded-full h-3 w-3 bg-green"></span>
      </span>
    </div>
    """
  end
```

File: lib/techschool_web/live/bootcamp_live/index.ex [hunk 2]
```diff
@@ -21,7 +28,13 @@ defmodule TechschoolWeb.BootcampLive.Index do
...
     )
+    |> assign(:online_users_count, 0)
     |> stream(:bootcamps, Bootcamps.list_bootcamps(), reset: true)
     |> ok()
   end
+
+  @impl true
+  def handle_info(%Phoenix.Socket.Broadcast{} = _event, socket) do
+    {:noreply, assign(socket, :online_users_count, OnlineUsersCounter.get_online_users_count())}
+  end
```
Comment (danielbergholz) [inline lib/techschool_web/live/bootcamp_live/index.ex]:
```suggestion
  def handle_info(%Phoenix.Socket.Broadcast{} = _event, socket) do
    socket
    |> assign(:online_users_count, OnlineUsersCounter.get_online_users_count())
    |> noreply()
  end
```

File: lib/techschool_web/live/bootcamp_live/show.ex [hunk 3]
```diff
@@ -30,6 +37,11 @@ defmodule TechschoolWeb.BootcampLive.Show do
     |> ok()
   end
 
+  @impl true
+  def handle_info(%Phoenix.Socket.Broadcast{} = _event, socket) do
+    {:noreply, assign(socket, :online_users_count, OnlineUsersCounter.get_online_users_count())}
+  end
```
Comment (danielbergholz) [inline lib/techschool_web/live/bootcamp_live/show.ex]:
```suggestion
  def handle_info(%Phoenix.Socket.Broadcast{} = _event, socket) do
    socket
    |> assign(:online_users_count, OnlineUsersCounter.get_online_users_count())
    |> noreply()
  end
```

File: lib/techschool_web/live/course_live/index.ex [hunk 4]
```diff
@@ -105,6 +113,11 @@ defmodule TechschoolWeb.CourseLive.Index do
     |> noreply()
   end
 
+  @impl true
+  def handle_info(%Phoenix.Socket.Broadcast{} = _event, socket) do
+    {:noreply, assign(socket, :online_users_count, OnlineUsersCounter.get_online_users_count())}
+  end
```
Comment (danielbergholz) [inline lib/techschool_web/live/course_live/index.ex]:
```suggestion
  def handle_info(%Phoenix.Socket.Broadcast{} = _event, socket) do
    socket
    |> assign(:online_users_count, OnlineUsersCounter.get_online_users_count())
    |> noreply()
  end
```

File: lib/techschool_web/live/page_live/home.ex [hunk 5]
```diff
@@ -17,11 +18,23 @@ defmodule TechschoolWeb.PageLive.Home do
...
     end)
     |> assign(:page_title, "TechSchool")
+    |> assign(:online_users_count, 0)
     |> ok()
   end
+
+  @impl true
+  def handle_info(%Phoenix.Socket.Broadcast{} = _event, socket) do
+    {:noreply, assign(socket, :online_users_count, OnlineUsersCounter.get_online_users_count())}
+  end
```
Comment (danielbergholz) [inline lib/techschool_web/live/page_live/home.ex]:
```suggestion
  def handle_info(%Phoenix.Socket.Broadcast{} = _event, socket) do
    socket
    |> assign(:online_users_count, OnlineUsersCounter.get_online_users_count())
    |> noreply()
  end
```

File: test/techschool_web/components/online_users_test.exs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      assert rendered_component_html =~ ""
+    end
+
+    test "renders online users count singular and plural form" do
+      assigns = %{count: 1}
+
+      rendered_component_html = render_component(&OnlineUsers.show_online_users/1, assigns)
+
+      assert rendered_component_html =~
+               ~s(<span id=\"online_users_count\" class=\"mt-2 relative inline-flex\">\n  <button type=\"button\" class=\"inline-flex items-center px-4 py-2 font-semibold leading-6 text-sm shadow rounded-md text-green bg-white dark:bg-slate-800 transition ease-in-out duration-150 cursor-not-allowed ring-1 ring-slate-900/10 dark:ring-slate-200/20\" disabled=\"\">\n    1 online user\n  </button>\n  <span class=\"flex absolute h-3 w-3 top-0 right-0 -mt-1 -mr-1\">\n    <span class=\"animate-ping absolute inline-flex h-full w-full rounded-full bg-green opacity-75\">\n    </span>\n    <span class=\"relative inline-flex rounded-full h-3 w-3 bg-green\"></span>\n  </span>\n</span>)
```
Comment (danielbergholz) [inline test/techschool_web/components/online_users_test.exs]: I had to update the HTML form the tests after I changed it

```suggestion
               ~s(<div id=\"online_users_count\" class=\"mt-2 relative inline-flex\" aria-label=\"Number of online users using TechSchool\">\n  <p class=\"px-4 py-2 font-semibold text-sm rounded-md text-green bg-slate-800\">\n    1 user online\n  </p>\n  <span class=\"flex absolute h-3 w-3 top-0 right-0 -mt-1 -mr-1\">\n    <span class=\"animate-ping absolute inline-flex h-full w-full rounded-full bg-green opacity-75\">\n    </span>\n    <span class=\"relative inline-flex rounded-full h-3 w-3 bg-green\"></span>\n  </span>\n</div>)
```

File: test/techschool_web/components/online_users_test.exs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+      assigns = %{count: 2}
+
+      rendered_component_html = render_component(&OnlineUsers.show_online_users/1, assigns)
+
+      assert rendered_component_html =~
+               ~s(<span id=\"online_users_count\" class=\"mt-2 relative inline-flex\">\n  <button type=\"button\" class=\"inline-flex items-center px-4 py-2 font-semibold leading-6 text-sm shadow rounded-md text-green bg-white dark:bg-slate-800 transition ease-in-out duration-150 cursor-not-allowed ring-1 ring-slate-900/10 dark:ring-slate-200/20\" disabled=\"\">\n    2 online users\n  </button>\n  <span class=\"flex absolute h-3 w-3 top-0 right-0 -mt-1 -mr-1\">\n    <span class=\"animate-ping absolute inline-flex h-full w-full rounded-full bg-green opacity-75\">\n    </span>\n    <span class=\"relative inline-flex rounded-full h-3 w-3 bg-green\"></span>\n  </span>\n</span>)
```
Comment (danielbergholz) [inline test/techschool_web/components/online_users_test.exs]:
```suggestion
               ~s(<div id=\"online_users_count\" class=\"mt-2 relative inline-flex\" aria-label=\"Number of online users using TechSchool\">\n  <p class=\"px-4 py-2 font-semibold text-sm rounded-md text-green bg-slate-800\">\n    2 users online\n  </p>\n  <span class=\"flex absolute h-3 w-3 top-0 right-0 -mt-1 -mr-1\">\n    <span class=\"animate-ping absolute inline-flex h-full w-full rounded-full bg-green opacity-75\">\n    </span>\n    <span class=\"relative inline-flex rounded-full h-3 w-3 bg-green\"></span>\n  </span>\n</div>)
```

**Review by danielbergholz — COMMENTED** [review]
This is fantastic work @lcezermf! 

I basically made 3 suggestions:
- Change the UI for the button a little bit
- Update the tests to use the new UI
- I have a `ok` and `noreply` helper functions defined inside `lib/techschool_web.ex` that we can use to clean up the `{:noreply, ...}` tuple

These suggestions impact a lot of files, so feel free to just click on `commit changes` inside this PR if you agree with my comments!

**Comment by lcezermf** [comment]
@danielbergholz all accepted, running locally it keeps working fine.

**Comment by danielbergholz** [comment]
Impressive work @lcezermf! Thanks for spending the time to implement such a cool feature 🔥

**Review by danielbergholz — APPROVED** [review]

**Comment by lcezermf** [comment]
Close #79

**Pull request merged**